src/Features/LanguageServer/Protocol/Handler/CodeActions/CodeActionResolveHandler.cs

```diff
@@ -0,0 +1,257 @@
+            string? clientName,
+            CancellationToken cancellationToken)
+        {
+            var data = ((JToken)codeAction.Data).ToObject<CodeActionResolveData>();
+            var document = SolutionProvider.GetDocument(data.TextDocument, clientName);
+            var codeActions = await CodeActionsHandler.GetCodeActionsAsync(
+                document,
+                _codeFixService,
+                _codeRefactoringService,
+                data.Range,
+                cancellationToken).ConfigureAwait(false);
+
+            if (codeActions == null || !codeActions.Any())
+            {
+                return codeAction;
+            }
+
+            var codeActionToResolve = GetCodeActionToResolve(data.DistinctTitle, codeActions);
+
+            // We didn't find a matching action, so just return the action without an edit or command.
+            if (codeActionToResolve == null)
+            {
+                return codeAction;
+            }
+
+            var operations = await codeActionToResolve.GetOperationsAsync(cancellationToken).ConfigureAwait(false);
+            if (operations.IsEmpty)
+            {
+                return codeAction;
+            }
+
+            // TO-DO:
+            // 1) We currently must execute code actions which add new documents on the server as commands,
+            // since there is no LSP support for adding documents yet. In the future, we should move these actions
+            // to execute on the client.
+            // 2) There is also a bug (same tracking item) where code actions that edit documents other than the
+            // one where the code action was invoked from do not work. We must temporarily execute these as commands
+            // as well.
+            // https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1147293/
+            var runAsCommand = false;
+
+            var applyChangesOperations = operations.Where(operation => operation is ApplyChangesOperation);
+            if (applyChangesOperations.Any())
+            {
+                var workspaceEdits = await ComputeWorkspaceEdits(applyChangesOperations, document!, cancellationToken).ConfigureAwait(false);
+                if (workspaceEdits.Any())
+                {
+                    codeAction.Edit = new LSP.WorkspaceEdit { DocumentChanges = workspaceEdits };
+                }
+                else
+                {
+                    // The workspace edit is something we don't currently support, like adding a new document.
+                    runAsCommand = true;
+                }
+            }
+
+            // Set up to run as command on the server instead of using WorkspaceEdits.
+            var commandOperations = operations.All(operation => !(operation is ApplyChangesOperation));
+            if (commandOperations || runAsCommand)
+            {
+                codeAction.Command = SetCommand(codeAction, data);
+            }
+
+            return codeAction;
```

Comment:
the data flow here is really weird to read through.  lots of little state variables affecting flow.  i would restructure to be more like:

```
if this must be a command
   do all the logic for treating this like a command here
else
  do all teh logic for treating as an edit here
done
```

trying to fork out and then rejoin your data flow is just making this confusing imo.